# BE : Dynamically solving the TSP


### 10/10/2018 - Josselin DELOSTE 

###### Question 1 - Data structure

In [91]:
# reader student

"""A simple reader for data file for TSP problem"""
import sys
import time

## Before testing the code you have to change the text file data paths
file_5 = "/Users/josselindeloste/Desktop/data/tsp5.txt"
file_10 = "/Users/josselindeloste/Desktop/data/tsp10.txt"
file_15 = "/Users/josselindeloste/Desktop/data/tsp15.txt"
file_20 = "/Users/josselindeloste/Desktop/data/tsp20.txt"
file_25 = "/Users/josselindeloste/Desktop/data/tsp25.txt"
with open(file_15) as data_file:
    nb = 0

    # first line is total number of vertices in file
    print("expected number of vertices : {0}\n".format(int(data_file.readline())))
    Coord =[]
    for line in data_file:
        print("reading point {0}: {1}".format(nb,
                                              tuple(float(x) for x in line.split())))
        Coord.append(tuple(float(x) for x in line.split()))
        nb = nb + 1

    print("\nnumber of vertices: {0}".format(nb))
    

expected number of vertices : 15

reading point 0: (20833.3333, 17100.0)
reading point 1: (20900.0, 17066.6667)
reading point 2: (21300.0, 13016.6667)
reading point 3: (21600.0, 14150.0)
reading point 4: (21600.0, 14966.6667)
reading point 5: (21600.0, 16500.0)
reading point 6: (22183.3333, 13133.3333)
reading point 7: (22583.3333, 14300.0)
reading point 8: (22683.3333, 12716.6667)
reading point 9: (23616.6667, 15866.6667)
reading point 10: (23700.0, 15933.3333)
reading point 11: (23883.3333, 14533.3333)
reading point 12: (24166.6667, 13250.0)
reading point 13: (25149.1667, 12365.8333)
reading point 14: (26133.3333, 14500.0)

number of vertices: 15


In [92]:
## setting up of our data structures in order to represent nodes and setofnodes
import networkx as nx
from scipy.spatial.distance import euclidean
import numpy as np
import matplotlib.pyplot as plt

D = np.zeros((len(Coord),len(Coord)))
for i in range(0,len(Coord)):
    for j in range(i+1,len(Coord)):
        D[i,j] = euclidean(Coord[i],Coord[j])
        D[j,i] = D[i,j]

# D is our matrix containing all euclidean distances beetween cities

print(D)

#g = nx.Graph()
#g = g_5.add_weighted_edges_from(zip([1,1,1,1,2,2,2,3,3,4],
 #                                     [2,3,4,5,3,4,5,4,5,5],
                                      
print(D)

[[   0.           74.53561416 4109.91345989 3047.99570684 2266.91173136
   973.53881735 4190.10079937 3301.89339622 4757.74219761 3044.34818055
  3094.97805449 3986.26114911 5092.5054301  6406.1495674  5903.38885726]
 [  74.53561416    0.         4069.70514902 2999.49072992 2213.59436212
   900.61709338 4137.39724881 3238.52676811 4701.3591289  2969.89527743
  3020.6695608  3913.82877684 5023.74920033 6336.65939584 5828.85540889]
 [4109.91345989 4069.70514902    0.         1172.36699411 1972.94196569
  3496.22809309  891.0043852  1814.90735791 1415.4896746  3672.79792514
  3777.16084587 2995.64492463 2876.14714467 3903.80178787 5055.82717938]
 [3047.99570684 2999.49072992 1172.36699411    0.          816.6667
  2350.         1172.13007716  994.70818781 1796.67898852 2648.37481067
  2755.04585423 2315.28732078 2719.88564997 3972.38405467 4546.82425533]
 [2266.91173136 2213.59436212 1972.94196569  816.6667        0.
  1533.3333     1923.89945019 1188.01888359 2497.22066283 2208.38053308


###### Question 2 - Held-Karp’s algorithm implementation

In [93]:
import itertools

tim = time.time()

# We use a dictionary :
L = {}
n = len(Coord)
for i in range(1, n):
    L[((0,i),i)] = D[i,0]
    
for m in range(2,n):
    for i in itertools.combinations(range(1,n), m):
        i = set(i)
        i.add(0)
        i = tuple(sorted(i))
        for j in i:
            if j != 0:
                s = set(i)
                s.remove(j)
                s = tuple(sorted(s))
                min_temp = float('inf')
                for k in i:
                    if k != j and k != 0:
                        value = L.get((s,k)) + D[k,j]
                        min_temp = min(value, min_temp)
                L[(i,j)] = min_temp

sol = float('inf')
for j in range(1,n):
    sol = min(sol, L.get((tuple(sorted(range(n))),j)) + D[j,0])

print("number of cities : "+str(n))
print("Optimal distance : "+str(sol))
print("Execution time : "+str(time.time()-tim))

## In order to test the algorithm with an increasing number of cities, please select
## the corresponding text file in first code window (file_5, file_10, file_15, file_20, file_25)

number of cities : 15
Optimal distance : 19269.935997794255
Execution time : 2.024384021759033


##### Some results :
- number of cities : 5 
    - Optimal distance : ~ 8387.08
    - Execution time : ~ 0.0024
- number of cities : 10 
    - Optimal distance : ~ 12349.98
    - Execution time : ~ 0.0404
- number of cities : 15 
    - Optimal distance : ~ 19269.94
    - Execution time : ~ 2.0244
- number of cities : 20 
    - Optimal distance : ~ 23328.99
    - Execution time : ~ 86.1478 (~ 1 min and 26 seconds)
- number of cities : 25 
    - Sorry we haven't suceeded to code our algorithm in less than minus 30 minutes :)